<a href="https://colab.research.google.com/github/PhoenixCC0722/Journey_to_become_DataScientist/blob/main/Chapter7_supervised_ML_competition2906_housing_iteration_5_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Competition of Supervised ML for the housing data**

*   Dataset: We'll be using the housing dataset available on our learning platform. Find it here 👉🏽 https://learn.wbscodingschool.com/courses/data-science/lessons/🏗-3-ml-classification/topic/📝-iteration-5-model-selection/.
*   Goal: Practice all day to craft the best scoring model.
*   Challenge: Once you've perfected your model, put it to the test with the unseen data attached to this post.
*   Submission: Should be done here https://bestpricehouse.streamlit.app/ Make sure you have the correct format for the submission .csv

### **Step 1. Data reading & splitting**

In [48]:
# required packages
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

>> How to get the url and the **correct path** of the .csv file

-- download the dataset from the Learning Platform

-- upload the .csv filt to Google Drive

-- find the file on the Google Drive website and the choose 'open with Google sheet'

!!! how to set the "share" --> anyone on the internet with the link can open it (bcz default is only me can read it)

In [2]:
# 1st way
url = "https://drive.google.com/file/d/19bkrhj__h0cEKgpjc-DXcNhF2A30UKhI/view?usp=sharing" # the .csv file is already uploaded from local server to the colad in google drive. In colab notebook, 'share' --> anyone with the link --> 'copy the link'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]

data = pd.read_csv(path)
data.columns

Index(['LotArea', 'LotFrontage', 'TotalBsmtSF', 'BedroomAbvGr', 'Fireplaces',
       'PoolArea', 'GarageCars', 'WoodDeckSF', 'ScreenPorch', 'Expensive',
       'MSZoning', 'Condition1', 'Heating', 'Street', 'CentralAir',
       'Foundation', 'ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'KitchenQual', 'FireplaceQu',
       'MSSubClass', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', '1stFlrSF',
       '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath',
       'FullBath', 'HalfBath', 'KitchenAbvGr', 'TotRmsAbvGrd', 'GarageYrBlt',
       'GarageArea', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'MiscVal',
       'MoSold', 'YrSold', 'Id', 'Alley', 'LotShape', 'LandContour',
       'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrTyp

In [3]:
# 2nd way
# Get the spreadsheet ID and name
# google sheets address: https://docs.google.com/spreadsheets/d/1Y5iy40vxJ8QCXpgz4O4UOUhJ3Fx6chj4Bkp0h4mJx_c/edit?gid=1175286409#gid=1175286409
sheet_id = "1Y5iy40vxJ8QCXpgz4O4UOUhJ3Fx6chj4Bkp0h4mJx_c"
sheet_name = "1175286409"

# Construct the URL to the CSV export of the spreadsheet
url = f"https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=csv&gid={sheet_name}"

# Read the CSV data into a Pandas DataFrame
data = pd.read_csv(url)
data.columns

Index(['LotArea', 'LotFrontage', 'TotalBsmtSF', 'BedroomAbvGr', 'Fireplaces',
       'PoolArea', 'GarageCars', 'WoodDeckSF', 'ScreenPorch', 'Expensive',
       'MSZoning', 'Condition1', 'Heating', 'Street', 'CentralAir',
       'Foundation', 'ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'KitchenQual', 'FireplaceQu',
       'MSSubClass', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', '1stFlrSF',
       '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath',
       'FullBath', 'HalfBath', 'KitchenAbvGr', 'TotRmsAbvGrd', 'GarageYrBlt',
       'GarageArea', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'MiscVal',
       'MoSold', 'YrSold', 'Id', 'Alley', 'LotShape', 'LandContour',
       'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrTyp

In [4]:
data.dtypes

LotArea            int64
LotFrontage      float64
TotalBsmtSF        int64
BedroomAbvGr       int64
Fireplaces         int64
                  ...   
PoolQC            object
Fence             object
MiscFeature       object
SaleType          object
SaleCondition     object
Length: 81, dtype: object

In [46]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   LotArea        1460 non-null   int64  
 1   LotFrontage    1201 non-null   float64
 2   TotalBsmtSF    1460 non-null   int64  
 3   BedroomAbvGr   1460 non-null   int64  
 4   Fireplaces     1460 non-null   int64  
 5   PoolArea       1460 non-null   int64  
 6   GarageCars     1460 non-null   int64  
 7   WoodDeckSF     1460 non-null   int64  
 8   ScreenPorch    1460 non-null   int64  
 9   MSZoning       1460 non-null   object 
 10  Condition1     1460 non-null   object 
 11  Heating        1460 non-null   object 
 12  Street         1460 non-null   object 
 13  CentralAir     1460 non-null   object 
 14  Foundation     1460 non-null   object 
 15  ExterQual      1460 non-null   object 
 16  ExterCond      1460 non-null   object 
 17  BsmtQual       1423 non-null   object 
 18  BsmtCond

In [6]:
# X (feature vectors) and y (label/target) creation
X = data
y = X.pop("Expensive")

# data splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

# building the pipeline
X_cat = X.select_dtypes(exclude="number").copy()
X_num = X.select_dtypes(include="number").copy()

numeric_pipe = make_pipeline(
    SimpleImputer(strategy="mean"))

categoric_pipe = make_pipeline(
    SimpleImputer(strategy="constant", fill_value="N_A"),
    OneHotEncoder(handle_unknown="ignore")
)

preprocessor = ColumnTransformer(
    transformers=[
        ("num_pipe", numeric_pipe, X_num.columns),
        ("cat_pipe", categoric_pipe, X_cat.columns),
    ]
)

In [12]:
X_test.tail(20)

LotArea  LotFrontage  TotalBsmtSF  BedroomAbvGr  Fireplaces  PoolArea  \
1203     9750         75.0         1630             3           1         0   
163      5500         55.0          882             1           0         0   
270     10728         84.0         1095             3           1         0   
33      10552         70.0         1398             4           1         0   
385      3182         43.0         1256             2           1         0   
929     13006          NaN          964             4           1         0   
966      9600        130.0          728             2           2         0   
427      8593         77.0          907             3           0         0   
873     12144         60.0          832             3           1         0   
486     10289         79.0         1073             3           0         0   
646      7200         60.0            0             3           0         0   
996     10659          NaN         1050             3           0         0   
623      2117          NaN          756             2           1         0   
1028     9492         79.0          768             3           1         0   
1108     8063          NaN          924             3           1         0   
1012    10592         55.0          602             3           2         0   
1059    11275          NaN          854             3           1         0   
890      8064         60.0          672             3           1         0   
628     11606         70.0         1040             5           2         0   
352      9084         60.0          616             3           0         0   

      GarageCars  WoodDeckSF  ScreenPorch MSZoning  ... GarageType  \
1203           2          74            0       RL  ...     Attchd   
163            0           0            0       RL  ...        NaN   
270            3         192            0       FV  ...     Attchd   
33             2           0            0       RL  ...     Attchd   
385            2         146          144       RL  ...     Attchd   
929            2           0            0       RL  ...    BuiltIn   
966            1           0            0       RL  ...     Detchd   
427            1           0            0       RL  ...     Detchd   
873            1           0            0       RL  ...     Attchd   
486            2           0            0       RL  ...     Attchd   
646            2           0            0       RL  ...     Detchd   
996            1           0            0       RL  ...     Attchd   
623            2           0            0       FV  ...     Detchd   
1028           1           0            0       RL  ...     Attchd   
1108           2         100            0       RL  ...     Attchd   
1012           1          96           53       RL  ...     Detchd   
1059           2           0            0       RL  ...     Detchd   
890            2           0            0       RL  ...     Detchd   
628            2         335            0       RL  ...     Attchd   
352            1          48            0       RL  ...     Detchd   

     GarageFinish GarageQual GarageCond PavedDrive PoolQC  Fence MiscFeature  \
1203          Unf         TA         TA          Y    NaN    NaN         NaN   
163           NaN        NaN        NaN          Y    NaN  MnPrv         NaN   
270           RFn         TA         TA          Y    NaN    NaN         NaN   
33            RFn         TA         TA          Y    NaN    NaN         NaN   
385           Fin         TA         TA          Y    NaN    NaN         NaN   
929           Fin         TA         TA          Y    NaN    NaN         NaN   
966           Unf         TA         TA          Y    NaN    NaN         NaN   
427           Unf         TA         TA          Y    NaN    NaN         NaN   
873           Unf         TA         TA          Y    NaN    NaN        Othr   
486           RFn         TA         TA          Y    NaN    NaN         Na

### **Step 2-1. Modelling (Decision Tree)**


In [8]:
from sklearn.model_selection import GridSearchCV

full_pipeline = make_pipeline(preprocessor,
                              DecisionTreeClassifier())

param_grid = {
    "columntransformer__num_pipe__simpleimputer__strategy":["mean", "median"],
    "decisiontreeclassifier__max_depth": range(2, 14, 2),
    "decisiontreeclassifier__min_samples_leaf": range(3, 12, 2)
}

model_dtree = GridSearchCV(full_pipeline,
                      param_grid,
                      cv=5,
                      verbose=1)

model_dtree.fit(X_train, y_train)

scores = {"dtree" : model_dtree.best_score_}

scores

Fitting 5 folds for each of 60 candidates, totalling 300 fits


{'dtree': 0.9289314405194233}

>> show the accuracy value of the 60 candidates and print the best fitting parameters

-- create anempty list to store the accuracy values

-- best_params

In [9]:
# Create a list to store the results
results = []

# Iterate through the parameter combinations
for params in model_dtree.cv_results_['params']:
    # Find the score for the corresponding combination
    score = model_dtree.cv_results_['mean_test_score'][model_dtree.cv_results_['params'].index(params)]

    # Append the combination and score to the results list
    results.append([params, score])

# Create a Pandas DataFrame from the results
results_df = pd.DataFrame(results, columns=['Parameters', 'Score'])

# Print the table
print(results_df)

# Get the best parameters
best_params = model_dtree.best_params_
print("\nBest KNeighborsClassifier Parameters:")
for param_name, param_value in best_params.items():
    print(f"{param_name}: {param_value}")

# Get the score of the best parameters
best_score = model_dtree.best_score_
print(f"\nBest Score: {best_score}")

                                           Parameters     Score
0   {'columntransformer__num_pipe__simpleimputer__...  0.917820
1   {'columntransformer__num_pipe__simpleimputer__...  0.917820
2   {'columntransformer__num_pipe__simpleimputer__...  0.917820
3   {'columntransformer__num_pipe__simpleimputer__...  0.917820
4   {'columntransformer__num_pipe__simpleimputer__...  0.917820
5   {'columntransformer__num_pipe__simpleimputer__...  0.922094
6   {'columntransformer__num_pipe__simpleimputer__...  0.924654
7   {'columntransformer__num_pipe__simpleimputer__...  0.919530
8   {'columntransformer__num_pipe__simpleimputer__...  0.921246
9   {'columntransformer__num_pipe__simpleimputer__...  0.927240
10  {'columntransformer__num_pipe__simpleimputer__...  0.923785
11  {'columntransformer__num_pipe__simpleimputer__...  0.923803
12  {'columntransformer__num_pipe__simpleimputer__...  0.925498
13  {'columntransformer__num_pipe__simpleimputer__...  0.922956
14  {'columntransformer__num_pipe__simpl

>> **conclusion:**

Best KNeighborsClassifier Parameters:

columntransformer__num_pipe__simpleimputer__strategy: median

decisiontreeclassifier__max_depth: 6

decisiontreeclassifier__min_samples_leaf: 5

Best Score: 0.9289314405194233

In [10]:
# Error analysos
# test accuracy
accuracy_score(model_dtree.predict(X_test), y_test)

0.9315068493150684

In [14]:
# save the X_test file as csv file to get the accuracy
# Save X_test to CSV in Colab

from google.colab import drive
from google.colab import files

X_test.to_csv('X_test_set_classification.csv', index=False)

# Upload the CSV to Google Drive
files.download('X_test_set_classification.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [16]:
# read the file from the download folder
X_test_dtree = pd.read_csv("X_test_set_classification.csv")
X_test_dtree.tail()

LotArea  LotFrontage  TotalBsmtSF  BedroomAbvGr  Fireplaces  PoolArea  \
287    10592         55.0          602             3           2         0   
288    11275          NaN          854             3           1         0   
289     8064         60.0          672             3           1         0   
290    11606         70.0         1040             5           2         0   
291     9084         60.0          616             3           0         0   

     GarageCars  WoodDeckSF  ScreenPorch MSZoning  ... GarageType  \
287           1          96           53       RL  ...     Detchd   
288           2           0            0       RL  ...     Detchd   
289           2           0            0       RL  ...     Detchd   
290           2         335            0       RL  ...     Attchd   
291           1          48            0       RL  ...     Detchd   

    GarageFinish GarageQual GarageCond PavedDrive PoolQC  Fence MiscFeature  \
287          Unf         TA         TA          Y    NaN    NaN         NaN   
288          Unf         TA         Fa          Y    NaN    NaN         NaN   
289          Unf         TA         TA          Y    NaN  MnPrv        Shed   
290          Unf         TA         TA          Y    NaN    NaN         NaN   
291          Unf         TA         Fa          Y    NaN    NaN         NaN   

    SaleType SaleCondition  
287       WD        Normal  
288       WD        Normal  
289       WD        Normal  
290       WD        Family  
291    ConLw        Normal  

[5 rows x 80 columns]

In [29]:
len(X_test)

292

In [17]:
pred = model_dtree.predict(X_test_dtree)
pred

array([0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0])

In [45]:
pred_1 = model_dtree.predict(X)
len(pred_1)

1460

In [43]:
# save results
predictions = pd.DataFrame({"Id": X["Id"], "Expensive": pred_1})
predictions.to_csv("predictions_dtree_1.csv", index=False)

In [47]:
# Upload the CSV to Google Drive
files.download('predictions_dtree_1.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [42]:
len(predictions)

1460

In [36]:
predictions.tail()

Id  Expensive
1455  1456          0
1456  1457          0
1457  1458          1
1458  1459          0
1459  1460          0

In [22]:
# save results
predictions = pd.DataFrame({"Id": X_test_dtree["Id"], "Expensive": pred})
predictions.to_csv("predictions_dtree.csv", index=False)

In [23]:
predictions

Id  Expensive
0     148          0
1     677          0
2    1305          0
3    1373          0
4    1428          0
..    ...        ...
287  1013          0
288  1060          0
289   891          0
290   629          0
291   353          0

[292 rows x 2 columns]

In [25]:
predictions.Expensive.value_counts()

Expensive
0    248
1     44
Name: count, dtype: int64

In [26]:
# Upload the CSV to Google Drive
files.download('predictions_dtree.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### **Step 2-2. Modelling (RandomForestClassifier)**